### ISSUE TITLE
Monotonic deterioration of policy between nets, beyond any reasonable value

### ISSUE THREAD

**Issue by Gilad-B** [LeelaChessZero/lc0#714]
![policy](https://user-images.githubusercontent.com/47300329/52183456-b8693080-2810-11e9-8d90-e42769a2416d.gif)

White to play in the above position. FEN is   6k1/1pn5/p4Pp1/4R2p/P2qP3/4B2P/1PP2PB1/R5K1 w - - 0 5
The obvious move is Bxd4. The policy of NN 32876 gives this move P=16.24% (which maybe fine for all I know), yet NN 32932 gives it as low as P=0.21% (!), and later nets don't seem much better.

I listed below the values I got for Bxd4 with every NN I tried (no cherry-picking), when analyzing this fen. You can see that P of Bxd4 monotonically decreases between 32876 and 32932 (and stays very low later).

| NN        | P of Bxd4          |
|:-----------------|:---------------:|
|32860         |   15.04% |
|32876        |   16.24%      |  
|32884        |   13.50%|
|32892        |   12.62%|
|32900        |  7.14%|
|32908        |   2.51%|
|32911        |   0.79%|
|32914        |   0.53%|
|32924        |  0.38%|
|32928        |  0.23%|
|32932        |  0.21%|
|32936        |   0.25%|
|32940        |   0.34%|
|32960        |  0.74%|
|32988        |  0.44%|
|32991        |   0.44%|

For perspective, the move with the highest P with 32860 is Bxd4 (P=15.04%), with 32908 it is Rb1 (P=11.65%), with 32940 it is Rb1 (P=19.40%), and with 32991 it is f2f3 (P=22.13%).

If it's any help, I attach my logfile (many "go infinite" on this fen, each time with a different NN): [GiladLog.txt](https://github.com/LeelaChessZero/lc0/files/2826057/GiladLog.txt)

**Comment by Gilad-B** [comment]
I happened to see another (even more striking) case of policy deterioration, within a different period of networks. It was reported by @sergiovieri on discord.  In the follwoing position (Black's turn) the obvious correct move is Rxa8:

![sergio](https://user-images.githubusercontent.com/47300329/52541521-ca157f80-2d9e-11e9-88ba-f2f024ab41e2.gif)

Please see the policies that Sergio Vieri reported: https://discordapp.com/channels/425419482568196106/425419999096733706/534921529092341761
The couple of things he said immediately afterwards may also be relevant.

**Comment by Gilad-B** [comment]
By the way, the gentle reader may recall the loss of lc0 in game 7 of  the current CCC5 final against Stockfish. This loss led to the (maybe unjustified) banishment of net 33000 by an angry mob.
Here is the position after White's (lc0) 20th move in that game
(FEN:      r3k2r/pp3p1p/2p1b3/2bp2p1/2P5/4n1P1/PP1NB2P/R1Q4K b kq - 6 20 )

![blundergame7](https://user-images.githubusercontent.com/47300329/52542045-858ce280-2da4-11e9-8c19-b55e6e3f4e57.gif)

I did a quick check with a few nets (at low node count): 
NN 32591 needed only 53 nodes to correctly see that Black can play 20-d5xc4 and has a winning advantage. 
32940 reached the same conclusions after a few thousand nodes.
With a similar number of nodes, 32960 thought the position is about equal, and 32988 thought White has a winning advantage.

**Comment by Gilad-B** [comment]
Admittedly, this last example was just a quick check.
But anyway, the two previous examples of policy deterioration (mine and Sergio Vieri's) seem insane.
I'm no authority on these matters, but I cannot imagine that they could have materialized without some bug being involved.

**Comment by sergiovieri** [comment]
seems like each run has it's own bugs, t30's multi queen and t40's queen promotion with take? (5n2/4P2k/8/8/8/2K3Q1/8/8 w - - 0 1) not sure if t40 will ever recover from this

**Comment by Eddh** [comment]
It seems these are mostly endgame bugs. It may not be that much of a surprise considering that endgames are the least trained positions. Most training games end in 3fold, win by resignation, or insuficient mating material. So the kind of endgames where weird behavior happen are very rarely seen in training, and thus don't matter much in the weight optimizatinon.

**Comment by RedDenver** [comment]
Also, the L2 regularization slowly causes all the weights to tend toward a value of zero, so that may be what you're seeing. Combined with what @Eddh said about infrequent training on any particular position, it may just be that the network has other information it's storing than this particular case.

BTW, does leela still win in these positions? Because she doesn't care about losing material or shortest mate as long as she thinks the probability of winning is higher on a different move.

**Comment by Naphthalin** [comment]
Is this still an issue after using `--policy-softmax-temp=1.2` in training? This seems very T30-specific and can probably be closed.

**Comment by mooskagh** [comment]
This is a bug regarding a very old run.
It would be nice to have it as a regression test for new runs, but otherwise it's outdated..
So closing.
